examples/scripts/Custom NS with VW learner.py

```diff
@@ -0,0 +1,53 @@
+"""
+This is an example script that creates and execuates an Experiment.
+This script requires that the matplotlib and vowpalwabbit packages be installed.
+"""
+
+import coba as cb
+from coba.learners import VowpalMediator
+from typing import List
+
+# this mediator splits the 'x' namespace into multiple namespaces
+class CustomVWMediator(VowpalMediator):
+    def __init__(self, namespaces: List):
+        self.namespaces = namespaces
+        super().__init__()
+
+    def transform_example(self, vw_shared, vw_uniques, labels):
+        DEFAULT_NS = 'x'
+        num_groups = min(len(self.namespaces), len(vw_shared[DEFAULT_NS]))
+        num_feat_each_group = len(vw_shared[DEFAULT_NS]) // num_groups
+
+        for k, v in vw_shared[DEFAULT_NS].items():
+            new_ns = min(int(k) // num_feat_each_group, num_groups-1)
+            vw_shared[self.namespaces[new_ns]] = vw_shared.get(self.namespaces[new_ns], {})
+            vw_shared[self.namespaces[new_ns]][k] = v
+        
+        del vw_shared[DEFAULT_NS]
+
+#First, we define the learners that we want to test
+learners = [ cb.VowpalLearner(args="-q :: --cb_explore_adf --epsilon 0.05", vw=CustomVWMediator(['b', 'c'])), cb.VowpalLearner(vw=CustomVWMediator(['b','c']))]
+
+#Next we create an environment we'd like to evaluate against
+environments = cb.Environments.from_linear_synthetic(1000, n_action_features=0).shuffle([1])
+
+#We then create and run our experiment from our environments and learners
+result = cb.Experiment(environments,learners).run()
+
+# this is None if we use shuffle with multiple values
```

Comment:
not sure if this is a bug

Reply:
yes

Reply:
Ok, I may not be understanding correctly but what I think you're observing is that when shuffling with multiple seeds coba will make multiple environments (one for each seed). Then during the experiment, because there are multiple environments, coba will create a copy of the original learner for each environment so that learning doesn't bleed over between environments.

I couldn't really think of a clear way to decide which of the copied learners to "return" so we don't "return" any (i.e., the original learner is uninitialized because we only learned on copies of the original learner). When there is only one environment there is no ambiguity so the original learner is used during the experiment thus making it initialized after the experiment runs.

Does that make sense? I know the behavior is kind of unintuitive, but it was the best solution I could come up with.

Reply:
I'm assuming this is a side effect of forking processes? If that's the case, as long as we can register some sort of function that ensures that each learner on each env gets the .finish() call

Reply:
No, this is separate from forking/multiprocessing. This is just a side effect of there being multiple environments. Copying also happens when running in multiprocessing but in that case we don't do it. It just happens implicitly when passing the learner to a background process.

You are absolutely correct though. This is a problem for the new finish stuff... I hadn't considered it. Let me think a bit tonight and try to come up with a good solution.

Reply:
There's a actually a cheeky way we can get around this... The interface isn't great but it isn't horrible either. How often do you see it being a use case where someone wants to run an experiment with multiple environments and get VW learners back? Do you think it is enough that it is possible though maybe not super easy?